midend/expandMirror.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "expandMirror.h"
+#include "frontends/p4/coreLibrary.h"
+
+
+namespace P4 {
+
+void ExpandMirror::explode(const IR::Expression* expression, IR::Vector<IR::Expression>* output) {
+    if (auto st = expression->type->to<IR::Type_Header>()) {
+        for (auto f : st->fields) {
+            auto e = new IR::Member(f->type, expression, f->name);
+            explode(e, output);
+        }
+    } else {
+        BUG_CHECK(!expression->type->is<IR::Type_StructLike>() && !expression->type->is<IR::Type_Stack>(),
+                  "%1%: unexpected type", expression->type);
+        output->push_back(expression);
+    }
+}
+
+void ExpandMirror::postorder(IR::MethodCallExpression* expr) {
+    auto mi = MethodInstance::resolve(expr, refMap, typeMap);
+    auto em = mi->to<P4::ExternMethod>();
+    if (!em) {
+      return;
+    }
+    if (em->originalExternType->name.name != "Mirror") {
```

Comment:
I don't think you should hardwire such constants in a mid-end pass. The mid-end passes are to a large degree architecture-independent. If you want to do something like this you should consider using a policy class. See for example the ValidateTableProperties pass.

Reply:
What is architecture dependent here? How is converting a one form of expression to other dependent on architecture?

Reply:
The word "Mirror" does not appear in any of the P4 specs.
So it must be architecture-dependent.

Reply:
I haven't muddled too much in the front-end and so, I will followup with @cc10512 to make sure I understand your concern(s).  End of the day, it is a function called "foo" which I am changing the argument from one form to other which are orthogonal.

Reply:
You should not hardwire constants in the compiler if they are not part of the language. The compiler should work for all legal input programs. "Mirror" is a constant which is not part of the input language.

Reply:
I get what you are saying. But, how does this make a legal program illegal?

Reply:
It is ok for midend passes to be architecture dependent (many existing passes are); what they should not be is target depedent.  That said, it is desirable for them to be as general as possible, so as to be usable for multiple architectures.  As structured, this pass is closely tied to PSA (to support the PSA mirror extern), so it would be good to generalize the transform with a policy that decides when to expand things -- in the PSA case this would select argument 1 of any method call to the Mirror extern, but it would be nice if it could easily specify which arguments of which methods of which extern.

Reply:
Do you mean the PSA clone functionality?  There is no extern named Mirror in PSA.

Reply:
I am taking a function call (in this case "mirror") which is an identifier with an argument list and  match on that identifier and change the argument list in an orthogonal form.  And, the objection seems to be that identifier is not in the LRM as a string.  And, instead of hard-coding it, if I pass it as an argument, it is all OK. 

End of the day, the success of a language depends on the number of users and how it helps them - not how everything checks out.  My objective is to make a fix and move on.

Reply:
@krishna-barefoot I understand that this process can be frustrating, and you may be under tight time pressure, but please realize that p4c is intended to be used by other P4 targets that have no function called "Mirror" in them at all, that users might want to name something that of their own.  Does that make sense?